internal/index/index.go

```diff
@@ -14,25 +14,44 @@ import (
 const (
 	// IPAddressPoolRefCombinedField is an index for the poolRef of an IPAddress.
 	IPAddressPoolRefCombinedField = "index.poolRef"
+
+	// IPAddressClaimPoolRefCombinedField is an index for the poolRef of an IPAddressClaim.
+	IPAddressClaimPoolRefCombinedField = "index.poolRef"
 )
 
 // SetupIndexes adds indexes to the cache of a Manager.
 func SetupIndexes(ctx context.Context, mgr manager.Manager) error {
-	return mgr.GetCache().IndexField(ctx, &ipamv1.IPAddress{},
+	err := mgr.GetCache().IndexField(ctx, &ipamv1.IPAddress{},
 		IPAddressPoolRefCombinedField,
 		ipAddressByCombinedPoolRef,
 	)
+	if err != nil {
+		return err
+	}
+
+	return mgr.GetCache().IndexField(ctx, &ipamv1.IPAddressClaim{},
+		IPAddressClaimPoolRefCombinedField,
+		ipAddressClaimByCombinedPoolRef,
+	)
 }
 
 func ipAddressByCombinedPoolRef(o client.Object) []string {
 	ip, ok := o.(*ipamv1.IPAddress)
 	if !ok {
 		panic(fmt.Sprintf("Expected an IPAddress but got a %T", o))
 	}
-	return []string{IPAddressPoolRefValue(ip.Spec.PoolRef)}
+	return []string{IPPoolRefValue(ip.Spec.PoolRef)}
+}
+
+func ipAddressClaimByCombinedPoolRef(o client.Object) []string {
+	ip, ok := o.(*ipamv1.IPAddressClaim)
+	if !ok {
+		panic(fmt.Sprintf("Expected an IPAddressClaim but got a %T", o))
+	}
+	return []string{IPPoolRefValue(ip.Spec.PoolRef)}
 }
 
-// IPAddressPoolRefValue turns a corev1.TypedLocalObjectReference to an indexable value.
-func IPAddressPoolRefValue(ref corev1.TypedLocalObjectReference) string {
+// IPPoolRefValue turns a corev1.TypedLocalObjectReference to an indexable value.
+func IPPoolRefValue(ref corev1.TypedLocalObjectReference) string {
 	return fmt.Sprintf("%s%s", ref.Kind, ref.Name)
```

Comment:
Should we include namespace here to properly handle pools with the same name in different namespaces? Probably just an efficiency thing though.

Reply:
Yeah it makes sense to reduce the possibility to reconcile claims that have the same pool name in different namespaces. We decided to go with the approach of using `client.InNamespace` option when calling list which should also ensure the list call only looks in the namespace of the pool.

Note: We did also try it with GlobalInClusterIPPool, which doesn't have a namespace and it does seem to be fine with empty namespace.